enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,350 @@
+
+    // KMSKeyServiceAccount is the service account being used for the
+    // encryption request for the given KMS key. If absent, the Compute
+    // Engine default service account is used.
+    // See https://cloud.google.com/compute/docs/access/service-accounts#compute_engine_service_account
+    // for details on the default service account.
+    // +optional
+    KMSKeyServiceAccount string `json:"kmsKeyServiceAccount,omitempty"`
+}
+
+// GCPPlatformSpec holds the desired state of the Google Cloud Platform infrastructure provider.
+// This only includes fields that can be modified in the cluster.
+type GCPPlatformSpec struct{
+  // EncryptionKey defines the user-managed encryption keys to be used by the default storage class to
+  // encrypt volumes.
+  // +optional
+  EncryptionKey *EncryptionKeyReference `json:"encryptionKey,omitempty"`
+}
+
+// GCPPlatformStatus holds the current status of the Google Cloud Platform infrastructure provider.
+type GCPPlatformStatus struct {
+    // [... existing fields...]
+
+    // EncryptionKey defines the user-managed encryption keys used by the default storage class to
+    // encrypt volumes.
+    // +optional
+    EncryptionKey *EncryptionKeyReference `json:"encryptionKey,omitempty"`
+}
+```
+
+#### Install Config
+
+The API extensions above correspond to types already in use in the Installer install config: 
+[AWS](https://github.com/openshift/installer/blob/master/pkg/types/aws/machinepool.go#L100), 
+[GCP](https://github.com/openshift/installer/blob/master/pkg/types/gcp/machinepools.go#L73), &
+[Azure](https://github.com/openshift/installer/blob/master/pkg/types/azure/disk.go#L36).
+
+### Implementation Details/Notes/Constraints
+
+The Installer will write the encryption keys to the cluster infrastructure object when users have defined
+encryption keys on the `defaultMachinePlatform` (`installconfig.platform.(aws|azure|gcp).defaultMachinePlatform`).
+
+1. User adds encryption keys to the `defaultMachinePlatform` in the install config.
+2. Installer checks if `defaultMachinePlatform` has encryption keys defined, sets them in cluster infrastructure object.
+3. Cluster infrastructure object is created during install.
+4. cluster-storage-operator/CSI driver monitors cluster infrastructure object and creates default storage class
+accordingly
+5. Should the API extension be more broad to encompass a general storage class, rather than specifically geared toward
+encryption?
+
+### Risks and Mitigations
+
+No risks foreseen. 
+
+### Drawbacks
+
+These tasks could be handled day-2 by a user. The largest drawback would be in terms of the added complexity required to
+implement this feature when that could be handled day 2 by a user. The complexity of this task is quite small and seems
+reasonable, so this drawback does not seem like a reason to block this enhancement.
+
+## Design Details
+
+### Open Questions 
+
+1. Should these API extensions go in both the spec & status as proposed?
```

Comment:
What's the motivation for being in both? Does it allow validation, what would it mean for them to be different, how would you report errors?

Reply:
The idea is that user input would go in the `spec`, and once the operator has reconciled the `status` would be updated. So at install time, the installer would populate the `spec` indicating we want the default storage class to use this key. The status would be empty until the default storage class was created by the CSI driver (at which time `status` would be populated).

One of the motivations was to allow the encryption keys for the default storage class to be mutable. So the status would show the current key, and users could update the spec.  Based on Jan's comment, though, we may not want to make this mutable: https://github.com/openshift/enhancements/pull/1163#discussion_r936641956

> Does it allow validation, what would it mean for them to be different, how would you report errors?

I would need to lean on the storage team for the exact details of validation, but I would expect the status field to be sync'd from the spec field either after validation or the successful creation of the default storage class.

I will be honest that I am not an expert in operators as most of my time spent working on the installer does not involve them first-hand: I was under the impression this was the standard pattern. Is there a more obvious pattern?

Reply:
Yeah this is a pattern we've used before where we need to have the value accepted by a controller before it's applied. Seems like we need to work out the rotation and how that works before we decide on the API, if the API is immutable we probably don't need this in both spec and status

Reply:
@jsafrane @bertinatto can you provide some guidance here? This enhancement proposes that the specified keys should be mutable, but I am questioning that after Jan's comment here: https://github.com/openshift/enhancements/pull/1163#discussion_r936641956

If this field is not mutable, that means the cluster is stuck with one default storage class for the lifespan of the cluster, right? WDYT?

Reply:
We fixed all issues upstream, so in 4.13 it's harmless if OCP temporarily deletes and re-creates a StorageClass. Infrastructure field can be mutable.

Question is, what validation should OCP do when user changes a key - we can check that the key exists, but we probably can't check that VMs have permissions to use the key. It would require new permissions in the cloud and users are very reluctant to add them.

In addition, where should we report the validation / SC update errors? We can send events to our operator(s) + mark our `ClusterOperator` accordingly. Would users expect an error field in Infrastructure to tell why its status was not updated?